File: LeetCode/Problems/Python/146. LRU Cache.py
class DLinkedNode:
    def __init__(self):
        self.key = 0
        self.val = 0
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        #hashmap + double linked list
        self.capacity = capacity
        self.size = 0
        self.cache = {}
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        
        self.head.next = self.tail
        self.tail.prev= self.head


    def get(self, key: int) -> int:
        node = self.cache.get(key, None)
        if not node:
            return -1
        #move the accessed node to head
        self._move_to_head(node)
        return node.val
        

    def put(self, key: int, value: int) -> None:
        
        node = self.cache.get(key)
        
        if not node:
            #add 
            node = DLinkedNode()
            node.val = value
            node.key = key
            self.cache[key] = node
            self._add_node(node)  
            self.size+=1
            
            #check if over capacity
            if self.size > self.capacity:
                to_remove = self._pop_tail()
                del self.cache[to_remove.key]
                self.size-=1
            
            
        else:
            #update the val
            node.val = value
            self._move_to_head(node)
            
    def _move_to_head(self, node):
        self._remove_node(node)
        self._add_node(node)
        
        
    def _add_node(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def _remove_node(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
        
    def _pop_tail(self):
        last = self.tail.prev
        self._remove_node(last)
        return last


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: LeetCode/Problems/C++/146. LRU Cache.cpp
struct DLinkedNode {
    int val;
    int key;
    DLinkedNode* prev;
    DLinkedNode* next;
    DLinkedNode(){
        val = 0;
        key = 0;
        prev = NULL;
        next = NULL;
    }
    
};
class LRUCache {
private:
    int _capacity;
    int _size;
    unordered_map<int, DLinkedNode*> cache;
    DLinkedNode* head;
    DLinkedNode* tail;
    
    
    void addNode(DLinkedNode* node){
        //handle node
        node->prev = head;
        node->next = head->next;
        //handle head
        head->next->prev = node;
        head->next = node;
    }
    
    void removeNode(DLinkedNode* node){
        //handle node
        node->prev->next = node->next;
        //handle tail
        node->next->prev = node->prev;
    }
    
    
    
public:
    LRUCache(int capacity) {
        _capacity = capacity;
        _size = 0;
        head = new DLinkedNode();
        tail = new DLinkedNode();
        head->next = tail;
        tail->prev = head;
        
    }
    
    int get(int key) {
        if(cache.find(key) == cache.end()){
            return -1;
        }
        DLinkedNode* node = cache[key];
        removeNode(node);
        addNode(node);
        
        return node->val;
        
    }
    
    void put(int key, int value) {
        if(cache.find(key) == cache.end()){
            DLinkedNode* node = new DLinkedNode();
            node->key = key;
            node->val = value;
            cache[key] = node;
            addNode(node);
            _size++;
            
            if(_size > _capacity){
                DLinkedNode* to_pop = tail->prev;
                removeNode(to_pop);
                cache.erase(to_pop->key);
                _size--;
            }
            
        }
        else{
            DLinkedNode* node = cache[key];
            node->val = value;
            removeNode(node);
            addNode(node);
        }
    }
    
    
};